### DISCUSSION TITLE
How to trigger on a template x-for being loaded?

### DISCUSSION THREAD

**Discussion by MarcGodard** [alpinejs/alpine#1035] [1. Help] [answered]
How to trigger on a template x-for being loaded?

I would like the toast to self close after 8 seconds.

I tried the following:

`x-init="setTimeout(() => { $store.toastComponent.remove(index) }, 8000)"`

`x-on:load="setTimeout(() => { $store.toastComponent.remove(index) }, 8000)"`

Full Code:

```html
<section x-data>
  <div class="toast-wrap">
    <template x-for="(toast, index) in $store.toastComponent.toasts" class="relative">

      <div 
          x-show.transition.duration.1000ms="toast"
          @click="$store.toastComponent.remove(index)"
          x-init="setTimeout(() => { $store.toastComponent.remove(index) }, 8000)"
          class="alert" x-bind:class="{ 'alert-info': toast.type === 'info', 'alert-warning': toast.type === 'warning' }" role="alert">
        <div class="flex">
            <p class="font-bold" x-text="toast.title"></p>
            <p class="text-sm" x-text="toast.message"></p>
          </div>
        </div>
      </div>

    </template>
  </div>
</section>
```

**Comment by KevinBatdorf** [comment]
You have two options, I think.

1. Add the timer to the function that fires it. Here's [one example](https://codepen.io/KevinBatdorf/pen/JjGKbMa) I put together and [another](https://codepen.io/KevinBatdorf/pen/QWyQqZb?editors=1010) with timer animations.
2. Add a child element `x-init` implementation helper. I [Tweeted about this](https://twitter.com/kevinbatdorf/status/1351058338629488640) the other day. Note that Caleb said this functionality will be in V3, so anyone reading this later on, this recommendation is for V2

I think either approach will work. Option 2 probably is a bit more declarative, but requires hacking Alpine a bit.

**Comment by MarcGodard** [comment]
@KevinBatdorf Thanks for your reply.

1: This way doesn't work unless I want to add an `id` to each toast, this would significantly complicate this little component.
2: I took a look at this, and looks like a lot of code to add for something I would guess there would be an event I could fire on. I am also new to alpine and not ready to mess with the core functionality yet.

**Reply by KevinBatdorf** [reply]
I would add the ID then. It's pretty trivial. A third option could be to do something like this:

```html
<div x-show.transition.duration.1000ms="toast">
  <div
    x-data <!-- <-- new -->
    @click="$store.toastComponent.remove(index)"
    x-init="setTimeout(() => { $store.toastComponent.remove(index) }, 8000)"
    class="flex alert" x-bind:class="{ 'alert-info': toast.type === 'info', 'alert-warning': toast.type === 'warning' }" role="alert"
  >
    <p class="font-bold" x-text="toast.title"></p>
    <p class="text-sm" x-text="toast.message"></p>
    </div> <!-- By the way, looks like you have an extra /div here -->
   </div>
</div>
```

You will very likely need to tweak that to match your implementation.

Adding the `x-data` will scope a new component and Alpine will treat it as a fresh component when inserted via the `x-for`, and thus will trigger the init.

**Comment by MarcGodard** [comment]
My actual code has the right div amounts, must of been when I copied it because I removed all the svgs.

When I add an x-data I get error that things are undefined. Is there a way to carry a variable into a nested x-data?

**Reply by KevinBatdorf** [reply]
There isn't an Alpine way, but you could add an attribute to the parent element, then use vanilla JS to check the attributes on the `parentNode`

**Comment by MarcGodard** [comment, marked as answer]
Ok got it working this way:

```html
<section x-data>
  <div class="toast-wrap">
    <template x-for="(toast, index) in $store.toastComponent.toasts" class="relative">

      <div 
          x-show.transition.duration.1000ms="toast"
          @click="$store.toastComponent.remove(index)"
          @remove-toast="$store.toastComponent.remove(index)"
          class="alert" x-bind:class="{ 'alert-info': toast.type === 'info', 'alert-warning': toast.type === 'warning' }" role="alert">
          <div x-data x-init="setTimeout(() => { $dispatch('remove-toast') }, 8000)"></div>
          <div class="flex">
            <div>
              <p class="font-bold" x-text="toast.title"></p>
              <p class="text-sm" x-text="toast.message"></p>
            </div>
          </div>
        </div>
        
    </template>
  </div>
</section>
```

**Reply by MarcGodard** [reply]
Again same problem, the id is required :(

**Reply by KevinBatdorf** [reply]
I don't think you need `.window` on the listener though, right? Unless you're removing it from elsewhere.

**Reply by KevinBatdorf** [reply]
Nothing expensive about adding an id. If you want something more unique I often use this:
```js
id = Date.now() + Math.floor(Math.random() * 1000000)
```

**Reply by MarcGodard** [reply]
Removing the `window` worked :) Thought of that right after my comment.

**Reply by SimoTod** [reply]
Aren't you going to have issue when you have multiple element to toast that enter the array at the same time without a unique id?
I.e. The first element is added to the toasted variable in store and start the first timeout for index 0, at 4 sec you add a new element and it starts the timeout for index 1, at 8 sec the first element is removed and it triggers a rerendering, because you are not using keys, the second element gets rebuild from scratch in pos 1. Note there is still a timeout due for 12 sec. At 10 sec, you add a new element in position 1. At 12sec, the pending timeout kicks in removing the element in position 1 after just 2 seconds.

I think the component needs to be redesigned a bit.

**Reply by MarcGodard** [reply]
Yeah it worked better this way, but I will need to add an id... lol I was hoping to avoid that.
